File: frontend/src/api/traceability.ts
import { apiClient } from "./client";

export interface LinkedTestCase {
  test_case_id: string;
  title: string;
  link_status: string;
  link_id: string;
  link_type: string;
  confidence_score?: number;
}

export interface RequirementCoverage {
  requirement_id: string;
  requirement_title: string;
  external_id?: string;
  linked_test_cases: LinkedTestCase[];
  coverage_status: string;
}

export interface OrphanTestCase {
  test_case_id: string;
  title: string;
  external_id?: string;
}

export interface TraceabilityMatrix {
  coverage_percentage: number;
  total_requirements: number;
  covered_requirements: number;
  uncovered_requirements: number;
  total_test_cases: number;
  orphan_test_cases: number;
  matrix: RequirementCoverage[];
  orphans: OrphanTestCase[];
}

export interface AlgorithmMetrics {
  algorithm: string;
  total_suggestions: number;
  accepted_suggestions: number;
  rejected_suggestions: number;
  pending_suggestions: number;
  acceptance_rate: number;
}

export interface Metrics {
  coverage_percentage: number;
  suggestion_acceptance_rate: number;
  total_requirements: number;
  total_test_cases: number;
  total_links: number;
  total_suggestions: number;
  accepted_suggestions: number;
  rejected_suggestions: number;
  pending_suggestions: number;
  manual_links: number;
  ai_suggested_links: number;
  algorithm_breakdown: AlgorithmMetrics[];
}

const traceabilityApi = {
  async getMatrix(): Promise<TraceabilityMatrix> {
    const response = await apiClient.get<TraceabilityMatrix>("/traceability-matrix");
    return response.data;
  },

  async getMetrics(): Promise<Metrics> {
    const response = await apiClient.get<Metrics>("/metrics");
    return response.data;
  },

  async exportMatrix(format: "csv" | "json"): Promise<Blob> {
    const response = await apiClient.get(`/traceability-matrix/export?format=${format}`, {
      responseType: "blob",
    });
    return response.data;
  },

  downloadExport(blob: Blob, filename: string): void {
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement("a");
    link.href = url;
    link.download = filename;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  },
};

export default traceabilityApi;

File: backend/app/schemas/traceability.py
"""Schemas for traceability matrix and metrics"""

from uuid import UUID

from pydantic import BaseModel, ConfigDict


class LinkedTestCase(BaseModel):
    """Test case linked to a requirement"""

    test_case_id: UUID
    title: str
    link_status: str  # 'accepted', 'pending', 'rejected'
    link_id: UUID
    link_type: str
    confidence_score: float | None = None

    model_config = ConfigDict(from_attributes=True)


class RequirementCoverage(BaseModel):
    """Coverage information for a single requirement"""

    requirement_id: UUID
    requirement_title: str
    external_id: str | None = None
    linked_test_cases: list[LinkedTestCase]
    coverage_status: str  # 'covered', 'partially_covered', 'uncovered'

    model_config = ConfigDict(from_attributes=True)


class OrphanTestCase(BaseModel):
    """Test case with no linked requirements"""

    test_case_id: UUID
    title: str
    external_id: str | None = None

    model_config = ConfigDict(from_attributes=True)


class TraceabilityMatrixResponse(BaseModel):
    """Complete traceability matrix response"""

    coverage_percentage: float
    total_requirements: int
    covered_requirements: int
    uncovered_requirements: int
    total_test_cases: int
    orphan_test_cases: int
    matrix: list[RequirementCoverage]
    orphans: list[OrphanTestCase]

    model_config = ConfigDict(from_attributes=True)


class AlgorithmMetrics(BaseModel):
    """Metrics for a specific suggestion algorithm"""

    algorithm: str
    total_suggestions: int
    accepted_suggestions: int
    rejected_suggestions: int
    pending_suggestions: int
    acceptance_rate: float


class MetricsResponse(BaseModel):
    """Metrics response"""

    coverage_percentage: float
    suggestion_acceptance_rate: float
    total_requirements: int
    total_test_cases: int
    total_links: int
    total_suggestions: int
    accepted_suggestions: int
    rejected_suggestions: int
    pending_suggestions: int
    manual_links: int
    ai_suggested_links: int
    algorithm_breakdown: list[AlgorithmMetrics]

    model_config = ConfigDict(from_attributes=True)
